File: game/Levels.lua
-- [ts]: Levels.ts
local ____lualib = require("lualib_bundle") -- 1
local __TS__Class = ____lualib.__TS__Class -- 1
local ____exports = {} -- 1
local ____Config = require("game.Config") -- 6
local Config = ____Config.Config -- 6
--- 构造一条敌人技能。
local function makeSkill(id, name, damage, preparesNext, cooldownActions, effects) -- 46
	return { -- 47
		id = id, -- 47
		name = name, -- 47
		damage = damage, -- 47
		effects = effects == nil and ({}) or effects, -- 47
		preparesNext = preparesNext, -- 47
		cooldownActions = cooldownActions -- 47
	} -- 47
end -- 46
--- 构造波次定义（技能列表必须非空：至少要有一条冷却为 0 的普通攻击，保证敌人永远能出手）。
local function makeWave(name, isElite, hp, attack, armor, actionTurns, actionSeconds, skills) -- 51
	return { -- 52
		name = name, -- 52
		isElite = isElite, -- 52
		hp = hp, -- 52
		attack = attack, -- 52
		armor = armor, -- 52
		actionTurns = actionTurns, -- 52
		actionSeconds = actionSeconds, -- 52
		skills = skills -- 52
	} -- 52
end -- 51
--- 封锁棋盘的效果条目（走统一执行器的棋盘效果，等级数据只给格数与时机）。
local function blockSpec(cells) -- 56
	return {kind = "boardBlock", value = cells, target = "board"} -- 57
end -- 56
--- 默认关卡：数值严格等于 Config.Waves（自检以字面量锁定该等式），
-- 精英技能 = 蓄力重击（伤害 = Config 的重击 24，效果 = 封锁棋盘）在前、普通攻击在后，
-- 因此 `new Combat(new Board())` 保持 M5 已验收的「先蓄力 → 再重击」时序。
local function buildDefaultLevel() -- 65
	local waves = {} -- 66
	do -- 66
		local i = 0 -- 67
		while i < #Config.Waves do -- 67
			local nums = Config.Waves[i + 1] -- 68
			local strike = makeSkill( -- 69
				"strike", -- 69
				"普通攻击", -- 69
				nums.attack, -- 69
				false, -- 69
				0 -- 69
			) -- 69
			if nums.isElite then -- 69
				local heavy = makeSkill( -- 71
					"heavy", -- 71
					"蓄力重击", -- 71
					nums.heavyAttack, -- 71
					true, -- 71
					0, -- 71
					{blockSpec(Config.EliteHeavyBlockCells)} -- 71
				) -- 71
				waves[#waves + 1] = makeWave( -- 72
					nums.name, -- 72
					nums.isElite, -- 72
					nums.hp, -- 72
					nums.attack, -- 72
					nums.armor, -- 72
					nums.actionTurns, -- 72
					nums.actionSeconds, -- 72
					{heavy, strike} -- 72
				) -- 72
			else -- 72
				waves[#waves + 1] = makeWave( -- 74
					nums.name, -- 74
					nums.isElite, -- 74
					nums.hp, -- 74
					nums.attack, -- 74
					nums.armor, -- 74
					nums.actionTurns, -- 74
					nums.actionSeconds, -- 74
					{strike} -- 74
				) -- 74
			end -- 74
			i = i + 1 -- 67
		end -- 67
	end -- 67
	return {id = 0, name = "默认关卡", subtitle = "数值与 Config.Waves 一致的兼容关卡（自检基准）", waves = waves} -- 77
end -- 65
--- 6 个手工关卡：逐关引入新机制 ——
-- 1 关仅普通攻击；2 关节奏加快；3 关首次出现棋盘封锁；4 关封锁常态化；
-- 5 关双精英；6 关高频高压（通关后解锁无尽挑战，复用本关波次表并按轮次放大）。
local function buildLevels() -- 90
	return { -- 91
		{ -- 92
			id = 1, -- 93
			name = "试炼场", -- 94
			subtitle = "敌人只会普通攻击，先熟悉连线与结算节奏", -- 95
			waves = { -- 96
				makeWave( -- 97
					"游荡者", -- 97
					false, -- 97
					36, -- 97
					7, -- 97
					3, -- 97
					3, -- 97
					5, -- 97
					{makeSkill( -- 97
						"strike", -- 97
						"普通攻击", -- 97
						7, -- 97
						false, -- 97
						0 -- 97
					)} -- 97
				), -- 97
				makeWave( -- 98
					"游荡者", -- 98
					false, -- 98
					54, -- 98
					9, -- 98
					5, -- 98
					3, -- 98
					4.5, -- 98
					{makeSkill( -- 98
						"strike", -- 98
						"普通攻击", -- 98
						9, -- 98
						false, -- 98
						0 -- 98
					)} -- 98
				), -- 98
				makeWave( -- 99
					"看门人", -- 99
					true, -- 99
					100, -- 99
					12, -- 99
					8, -- 99
					3, -- 99
					4, -- 99
					{makeSkill( -- 99
						"strike", -- 99
						"普通攻击", -- 99
						12, -- 99
						false, -- 99
						0 -- 99
					)} -- 99
				) -- 99
			} -- 99
		}, -- 99
		{ -- 102
			id = 2, -- 103
			name = "锈蚀回廊", -- 104
			subtitle = "敌人出手更快，需要压缩每次决策时间", -- 105
			waves = { -- 106
				makeWave( -- 107
					"锈蚀兵", -- 107
					false, -- 107
					46, -- 107
					8, -- 107
					4, -- 107
					3, -- 107
					4.5, -- 107
					{makeSkill( -- 107
						"strike", -- 107
						"普通攻击", -- 107
						8, -- 107
						false, -- 107
						0 -- 107
					)} -- 107
				), -- 107
				makeWave( -- 108
					"锈蚀兵", -- 108
					false, -- 108
					68, -- 108
					11, -- 108
					6, -- 108
					2, -- 108
					4, -- 108
					{makeSkill( -- 108
						"strike", -- 108
						"普通攻击", -- 108
						11, -- 108
						false, -- 108
						0 -- 108
					)} -- 108
				), -- 108
				makeWave( -- 109
					"锈蚀队长", -- 109
					true, -- 109
					130, -- 109
					15, -- 109
					10, -- 109
					2, -- 109
					3.5, -- 109
					{ -- 109
						makeSkill( -- 109
							"heavy", -- 109
							"蓄力重击", -- 109
							26, -- 109
							true, -- 109
							0 -- 109
						), -- 109
						makeSkill( -- 109
							"strike", -- 109
							"普通攻击", -- 109
							15, -- 109
							false, -- 109
							0 -- 109
						) -- 109
					} -- 109
				) -- 109
			} -- 109
		}, -- 109
		{ -- 112
			id = 3, -- 113
			name = "封锁者", -- 114
			subtitle = "首次出现棋盘封锁：被锁的格子无法连线", -- 115
			waves = { -- 116
				makeWave( -- 117
					"封锁学徒", -- 117
					false, -- 117
					50, -- 117
					9, -- 117
					4, -- 117
					3, -- 117
					4.5, -- 117
					{makeSkill( -- 117
						"strike", -- 117
						"普通攻击", -- 117
						9, -- 117
						false, -- 117
						0 -- 117
					)} -- 117
				), -- 117
				makeWave( -- 118
					"封锁者", -- 118
					false, -- 118
					74, -- 118
					12, -- 118
					6, -- 118
					3, -- 118
					4, -- 118
					{ -- 118
						makeSkill( -- 119
							"lockPunch", -- 119
							"封锁击", -- 119
							14, -- 119
							true, -- 119
							0, -- 119
							{blockSpec(1)} -- 119
						), -- 119
						makeSkill( -- 120
							"strike", -- 120
							"普通攻击", -- 120
							12, -- 120
							false, -- 120
							0 -- 120
						) -- 120
					} -- 120
				), -- 120
				makeWave( -- 122
					"封锁监工", -- 122
					true, -- 122
					145, -- 122
					16, -- 122
					10, -- 122
					2, -- 122
					3.5, -- 122
					{ -- 122
						makeSkill( -- 123
							"heavy", -- 123
							"蓄力封锁", -- 123
							28, -- 123
							true, -- 123
							0, -- 123
							{blockSpec(2)} -- 123
						), -- 123
						makeSkill( -- 124
							"strike", -- 124
							"普通攻击", -- 124
							16, -- 124
							false, -- 124
							0 -- 124
						) -- 124
					} -- 124
				) -- 124
			} -- 124
		}, -- 124
		{ -- 128
			id = 4, -- 129
			name = "铁壁工事", -- 130
			subtitle = "封锁常态化，场上会长期残留在障碍", -- 131
			waves = { -- 132
				makeWave( -- 133
					"工事兵", -- 133
					false, -- 133
					58, -- 133
					10, -- 133
					5, -- 133
					3, -- 133
					4.5, -- 133
					{ -- 133
						makeSkill( -- 134
							"lockPunch", -- 134
							"封锁击", -- 134
							12, -- 134
							true, -- 134
							1, -- 134
							{blockSpec(1)} -- 134
						), -- 134
						makeSkill( -- 135
							"strike", -- 135
							"普通攻击", -- 135
							10, -- 135
							false, -- 135
							0 -- 135
						) -- 135
					} -- 135
				), -- 135
				makeWave( -- 137
					"工事官", -- 137
					false, -- 137
					82, -- 137
					13, -- 137
					7, -- 137
					3, -- 137
					4, -- 137
					{ -- 137
						makeSkill( -- 138
							"lockPunch", -- 138
							"封锁击", -- 138
							15, -- 138
							true, -- 138
							1, -- 138
							{blockSpec(2)} -- 138
						), -- 138
						makeSkill( -- 139
							"strike", -- 139
							"普通攻击", -- 139
							13, -- 139
							false, -- 139
							0 -- 139
						) -- 139
					} -- 139
				), -- 139
				makeWave( -- 141
					"铁壁统领", -- 141
					true, -- 141
					165, -- 141
					18, -- 141
					12, -- 141
					2, -- 141
					3.5, -- 141
					{ -- 141
						makeSkill( -- 142
							"heavy", -- 142
							"蓄力封锁", -- 142
							32, -- 142
							true, -- 142
							0, -- 142
							{blockSpec(3)} -- 142
						), -- 142
						makeSkill( -- 143
							"strike", -- 143
							"普通攻击", -- 143
							18, -- 143
							false, -- 143
							0 -- 143
						) -- 143
					} -- 143
				) -- 143
			} -- 143
		}, -- 143
		{ -- 147
			id = 5, -- 148
			name = "双生精英", -- 149
			subtitle = "连续两波精英，蓄力重击接踵而至", -- 150
			waves = { -- 151
				makeWave( -- 152
					"双子卫兵", -- 152
					false, -- 152
					62, -- 152
					11, -- 152
					5, -- 152
					3, -- 152
					4, -- 152
					{makeSkill( -- 152
						"strike", -- 152
						"普通攻击", -- 152
						11, -- 152
						false, -- 152
						0 -- 152
					)} -- 152
				), -- 152
				makeWave( -- 153
					"左精英", -- 153
					true, -- 153
					140, -- 153
					16, -- 153
					10, -- 153
					2, -- 153
					3.5, -- 153
					{ -- 153
						makeSkill( -- 154
							"heavy", -- 154
							"蓄力重击", -- 154
							28, -- 154
							true, -- 154
							0 -- 154
						), -- 154
						makeSkill( -- 155
							"strike", -- 155
							"普通攻击", -- 155
							16, -- 155
							false, -- 155
							0 -- 155
						) -- 155
					} -- 155
				), -- 155
				makeWave( -- 157
					"右精英", -- 157
					true, -- 157
					175, -- 157
					20, -- 157
					12, -- 157
					2, -- 157
					3, -- 157
					{ -- 157
						makeSkill( -- 158
							"heavy", -- 158
							"蓄力封锁", -- 158
							34, -- 158
							true, -- 158
							0, -- 158
							{blockSpec(2)} -- 158
						), -- 158
						makeSkill( -- 159
							"strike", -- 159
							"普通攻击", -- 159
							20, -- 159
							false, -- 159
							0 -- 159
						) -- 159
					} -- 159
				) -- 159
			} -- 159
		}, -- 159
		{ -- 163
			id = 6, -- 164
			name = "终局回响", -- 165
			subtitle = "高频出手 + 大范围封锁，通关后解锁无尽挑战", -- 166
			waves = { -- 167
				makeWave( -- 168
					"回响残兵", -- 168
					false, -- 168
					70, -- 168
					13, -- 168
					6, -- 168
					2, -- 168
					4, -- 168
					{makeSkill( -- 168
						"strike", -- 168
						"普通攻击", -- 168
						13, -- 168
						false, -- 168
						0 -- 168
					)} -- 168
				), -- 168
				makeWave( -- 169
					"回响官", -- 169
					true, -- 169
					160, -- 169
					19, -- 169
					11, -- 169
					2, -- 169
					3.5, -- 169
					{ -- 169
						makeSkill( -- 170
							"lockPunch", -- 170
							"封锁击", -- 170
							24, -- 170
							true, -- 170
							1, -- 170
							{blockSpec(2)} -- 170
						), -- 170
						makeSkill( -- 171
							"strike", -- 171
							"普通攻击", -- 171
							19, -- 171
							false, -- 171
							0 -- 171
						) -- 171
					} -- 171
				), -- 171
				makeWave( -- 173
					"回响之主", -- 173
					true, -- 173
					210, -- 173
					23, -- 173
					14, -- 173
					2, -- 173
					3, -- 173
					{ -- 173
						makeSkill( -- 174
							"heavy", -- 174
							"蓄力封锁", -- 174
							40, -- 174
							true, -- 174
							0, -- 174
							{blockSpec(3)} -- 174
						), -- 174
						makeSkill( -- 175
							"strike", -- 175
							"普通攻击", -- 175
							23, -- 175
							false, -- 175
							0 -- 175
						) -- 175
					} -- 175
				) -- 175
			} -- 175
		} -- 175
	} -- 175
end -- 90
--- 关卡注册表：查表即可，不含任何按关卡/按技能的分支逻辑。
____exports.Levels = __TS__Class() -- 183
local Levels = ____exports.Levels -- 183
Levels.name = "Levels" -- 183
function Levels.prototype.____constructor(self) -- 183
end -- 183
function Levels.get(self, id) -- 192
	for ____, level in ipairs(____exports.Levels.List) do -- 193
		if level.id == id then -- 193
			return level -- 195
		end -- 195
	end -- 195
	return ____exports.Levels.List[1] -- 198
end -- 192
function Levels.waveOf(self, level, index) -- 202
	local list = level.waves -- 203
	if index < 0 or index >= #list then -- 203
		return list[#list] -- 205
	end -- 205
	return list[index + 1] -- 207
end -- 202
function Levels.skillOf(self, wave, index) -- 211
	local list = wave.skills -- 212
	if index < 0 or index >= #list then -- 212
		return ____exports.Levels:fallbackSkill(wave) -- 214
	end -- 214
	return list[index + 1] -- 216
end -- 211
function Levels.fallbackSkill(self, wave) -- 223
	return makeSkill( -- 224
		"strike", -- 224
		"普通攻击", -- 224
		wave.attack, -- 224
		false, -- 224
		0 -- 224
	) -- 224
end -- 223
Levels.Default = buildDefaultLevel() -- 223
Levels.List = buildLevels() -- 223
Levels.EndlessBase = ____exports.Levels.List[#____exports.Levels.List] -- 223
return ____exports -- 223

File: game/Levels.ts
// 关卡注册表：纯数据。波次表、敌人技能表与 6 个手工关卡全部写在这里。
// 敌人技能的效果同样交给 game/Effects.ts 的统一执行器，Combat 不写任何按技能/按 kind 的特例分支。
// 敌人伤害不进效果三元组（执行器没有难度与关卡上下文的缩放能力），而是技能自带的 damage 字段，
// 由 Combat 统一乘上「难度 × 关卡」缩放后再结算。

import { Config } from 'game/Config';
import { EffectKind, EffectSpec, EffectTarget } from 'game/Effects';

/**
 * 敌人技能：一次敌人行动的纯数据定义。
 * - damage：结算给玩家的基础伤害（0 表示本次出手不造成伤害，例如蓄力预告）。
 * - effects：经统一执行器下发的效果三元组（例如重击封锁棋盘）。
 * - preparesNext：true = 本次出手只是蓄力预告，下一次行动才真正释放该技能。
 * - cooldownActions：释放后需要经过多少次敌人行动才能再次被选中。
 */
export interface EnemySkillDef {
	id: string;
	name: string;
	damage: number;
	effects: EffectSpec[];
	preparesNext: boolean;
	cooldownActions: number;
}

/** 单波敌人的纯数据定义（不含技能；技能见 EnemySkillDef）。 */
export interface WaveDef {
	name: string;
	isElite: boolean;
	hp: number;
	attack: number;
	armor: number;
	actionTurns: number;
	actionSeconds: number;
	skills: EnemySkillDef[];
}

/** 一个手工关卡的纯数据定义。 */
export interface LevelDef {
	id: number;
	name: string;
	subtitle: string;
	waves: WaveDef[];
}

/** 构造一条敌人技能。 */
function makeSkill(id: string, name: string, damage: number, preparesNext: boolean, cooldownActions: number, effects?: EffectSpec[]): EnemySkillDef {
	return { id, name, damage, effects: effects === undefined ? [] : effects, preparesNext, cooldownActions };
}

/** 构造波次定义（技能列表必须非空：至少要有一条冷却为 0 的普通攻击，保证敌人永远能出手）。 */
function makeWave(name: string, isElite: boolean, hp: number, attack: number, armor: number, actionTurns: number, actionSeconds: number, skills: EnemySkillDef[]): WaveDef {
	return { name, isElite, hp, attack, armor, actionTurns, actionSeconds, skills };
}

/** 封锁棋盘的效果条目（走统一执行器的棋盘效果，等级数据只给格数与时机）。 */
function blockSpec(cells: number): EffectSpec {
	return { kind: EffectKind.BoardBlock, value: cells, target: EffectTarget.Board };
}

/**
 * 默认关卡：数值严格等于 Config.Waves（自检以字面量锁定该等式），
 * 精英技能 = 蓄力重击（伤害 = Config 的重击 24，效果 = 封锁棋盘）在前、普通攻击在后，
 * 因此 `new Combat(new Board())` 保持 M5 已验收的「先蓄力 → 再重击」时序。
 */
function buildDefaultLevel(): LevelDef {
	const waves: WaveDef[] = [];
	for (let i = 0; i < Config.Waves.length; i++) {
		const nums = Config.Waves[i];
		const strike = makeSkill('strike', '普通攻击', nums.attack, false, 0);
		if (nums.isElite) {
			const heavy = makeSkill('heavy', '蓄力重击', nums.heavyAttack, true, 0, [blockSpec(Config.EliteHeavyBlockCells)]);
			waves.push(makeWave(nums.name, nums.isElite, nums.hp, nums.attack, nums.armor, nums.actionTurns, nums.actionSeconds, [heavy, strike]));
		} else {
			waves.push(makeWave(nums.name, nums.isElite, nums.hp, nums.attack, nums.armor, nums.actionTurns, nums.actionSeconds, [strike]));
		}
	}
	return {
		id: 0,
		name: '默认关卡',
		subtitle: '数值与 Config.Waves 一致的兼容关卡（自检基准）',
		waves,
	};
}

/**
 * 6 个手工关卡：逐关引入新机制 ——
 * 1 关仅普通攻击；2 关节奏加快；3 关首次出现棋盘封锁；4 关封锁常态化；
 * 5 关双精英；6 关高频高压（通关后解锁无尽挑战，复用本关波次表并按轮次放大）。
 */
function buildLevels(): LevelDef[] {
	return [
		{
			id: 1,
			name: '试炼场',
			subtitle: '敌人只会普通攻击，先熟悉连线与结算节奏',
			waves: [
				makeWave('游荡者', false, 36, 7, 3, 3, 5, [makeSkill('strike', '普通攻击', 7, false, 0)]),
				makeWave('游荡者', false, 54, 9, 5, 3, 4.5, [makeSkill('strike', '普通攻击', 9, false, 0)]),
				makeWave('看门人', true, 100, 12, 8, 3, 4, [makeSkill('strike', '普通攻击', 12, false, 0)]),
			],
		},
		{
			id: 2,
			name: '锈蚀回廊',
			subtitle: '敌人出手更快，需要压缩每次决策时间',
			waves: [
				makeWave('锈蚀兵', false, 46, 8, 4, 3, 4.5, [makeSkill('strike', '普通攻击', 8, false, 0)]),
				makeWave('锈蚀兵', false, 68, 11, 6, 2, 4, [makeSkill('strike', '普通攻击', 11, false, 0)]),
				makeWave('锈蚀队长', true, 130, 15, 10, 2, 3.5, [makeSkill('heavy', '蓄力重击', 26, true, 0), makeSkill('strike', '普通攻击', 15, false, 0)]),
			],
		},
		{
			id: 3,
			name: '封锁者',
			subtitle: '首次出现棋盘封锁：被锁的格子无法连线',
			waves: [
				makeWave('封锁学徒', false, 50, 9, 4, 3, 4.5, [makeSkill('strike', '普通攻击', 9, false, 0)]),
				makeWave('封锁者', false, 74, 12, 6, 3, 4, [
					makeSkill('lockPunch', '封锁击', 14, true, 0, [blockSpec(1)]),
					makeSkill('strike', '普通攻击', 12, false, 0),
				]),
				makeWave('封锁监工', true, 145, 16, 10, 2, 3.5, [
					makeSkill('heavy', '蓄力封锁', 28, true, 0, [blockSpec(2)]),
					makeSkill('strike', '普通攻击', 16, false, 0),
				]),
			],
		},
		{
			id: 4,
			name: '铁壁工事',
			subtitle: '封锁常态化，场上会长期残留在障碍',
			waves: [
				makeWave('工事兵', false, 58, 10, 5, 3, 4.5, [
					makeSkill('lockPunch', '封锁击', 12, true, 1, [blockSpec(1)]),
					makeSkill('strike', '普通攻击', 10, false, 0),
				]),
				makeWave('工事官', false, 82, 13, 7, 3, 4, [
					makeSkill('lockPunch', '封锁击', 15, true, 1, [blockSpec(2)]),
					makeSkill('strike', '普通攻击', 13, false, 0),
				]),
				makeWave('铁壁统领', true, 165, 18, 12, 2, 3.5, [
					makeSkill('heavy', '蓄力封锁', 32, true, 0, [blockSpec(3)]),
					makeSkill('strike', '普通攻击', 18, false, 0),
				]),
			],
		},
		{
			id: 5,
			name: '双生精英',
			subtitle: '连续两波精英，蓄力重击接踵而至',
			waves: [
				makeWave('双子卫兵', false, 62, 11, 5, 3, 4, [makeSkill('strike', '普通攻击', 11, false, 0)]),
				makeWave('左精英', true, 140, 16, 10, 2, 3.5, [
					makeSkill('heavy', '蓄力重击', 28, true, 0),
					makeSkill('strike', '普通攻击', 16, false, 0),
				]),
				makeWave('右精英', true, 175, 20, 12, 2, 3, [
					makeSkill('heavy', '蓄力封锁', 34, true, 0, [blockSpec(2)]),
					makeSkill('strike', '普通攻击', 20, false, 0),
				]),
			],
		},
		{
			id: 6,
			name: '终局回响',
			subtitle: '高频出手 + 大范围封锁，通关后解锁无尽挑战',
			waves: [
				makeWave('回响残兵', false, 70, 13, 6, 2, 4, [makeSkill('strike', '普通攻击', 13, false, 0)]),
				makeWave('回响官', true, 160, 19, 11, 2, 3.5, [
					makeSkill('lockPunch', '封锁击', 24, true, 1, [blockSpec(2)]),
					makeSkill('strike', '普通攻击', 19, false, 0),
				]),
				makeWave('回响之主', true, 210, 23, 14, 2, 3, [
					makeSkill('heavy', '蓄力封锁', 40, true, 0, [blockSpec(3)]),
					makeSkill('strike', '普通攻击', 23, false, 0),
				]),
			],
		},
	];
}

/** 关卡注册表：查表即可，不含任何按关卡/按技能的分支逻辑。 */
export class Levels {
	/** 兼容关卡：数值与 Config.Waves 完全一致，供未指定关卡的 Combat 使用。 */
	static readonly Default: LevelDef = buildDefaultLevel();
	/** 6 个手工关卡（顺序解锁，id 从 1 开始）。 */
	static readonly List: LevelDef[] = buildLevels();
	/** 无尽挑战的基础关卡：复用最后一关的波次表，轮次由关卡缩放实现。 */
	static readonly EndlessBase: LevelDef = Levels.List[Levels.List.length - 1];

	/** 按 id 取关卡；id 越界时回退到第 1 关。 */
	static get(id: number): LevelDef {
		for (const level of Levels.List) {
			if (level.id === id) {
				return level;
			}
		}
		return Levels.List[0];
	}

	/** 取指定波次；越界时回退到最后一波（与 Config.wave 的容错语义一致）。 */
	static waveOf(level: LevelDef, index: number): WaveDef {
		const list = level.waves;
		if (index < 0 || index >= list.length) {
			return list[list.length - 1];
		}
		return list[index];
	}

	/** 取指定技能；下标越界时回退到本波自身的普通攻击（与 fallbackSkill 语义一致）。 */
	static skillOf(wave: WaveDef, index: number): EnemySkillDef {
		const list = wave.skills;
		if (index < 0 || index >= list.length) {
			return Levels.fallbackSkill(wave);
		}
		return list[index];
	}

	/**
	 * 波次数据兜底：该波没有任何技能（或技能全在冷却）时使用的普通攻击。
	 * 数值取自波次自身的 attack，因此依然是纯数据推导，不引入按 kind 的分支。
	 */
	static fallbackSkill(wave: WaveDef): EnemySkillDef {
		return makeSkill('strike', '普通攻击', wave.attack, false, 0);
	}
}